src/main/webapp/app/exam/monitor/exam-monitoring.service.ts

```diff
@@ -0,0 +1,70 @@
+import { Exam } from 'app/entities/exam.model';
+
+@Injectable({ providedIn: 'root' })
+export class ExamMonitoringService {
+    constructor(private serverDateService: ArtemisServerDateService, private http: HttpClient) {}
+
+    /**
+     * Receives the event and adds a timestamp.
+     * @param studentExam current student exam
+     * @param examAction performed action
+     * @param monitoring true if monitoring is enabled
+     */
+    public handleActionEvent(studentExam: StudentExam, examAction: ExamAction, monitoring: boolean) {
+        if (!monitoring) {
+            return;
+        }
+        const examActivity = studentExam.examActivity || new ExamActivity();
+        examAction.timestamp = this.serverDateService.now();
+        examActivity.addAction(examAction);
+        studentExam.examActivity = examActivity;
+    }
+
+    /**
+     * Saves the actions to the server and removes them from the activity. If the save operation wasn't successful, the actions are not removed.
+     * @param exam current exam
+     * @param studentExam current student exam
+     * @param courseId which the exam belongs to
+     */
+    public saveActions(exam: Exam, studentExam: StudentExam, courseId: number) {
+        // We synchronize the user actions with the server and then delete them on the client, as they are no longer used
+        if (exam.monitoring && studentExam.examActivity) {
+            const actionsToSync = studentExam.examActivity.examActions;
+            this.syncActions(actionsToSync, courseId, exam.id!, studentExam.id!).subscribe({
+                // After successful synchronization we can delete the actions -> filter in case of new actions during the synchronization
+                next: () => (studentExam.examActivity!.examActions = studentExam.examActivity!.examActions.filter((action) => !actionsToSync.includes(action))),
+                // We do not delete the client actions, because they are not synchronized yet
+                error: () => {},
+            });
+        }
+    }
+
+    /**
+     * Returns the resource url.
+     * @param courseId which the exam belongs to
+     * @param examId of the current exam
+     * @private
+     */
+    public getResourceURL(courseId: number, examId: number): string {
+        return `${SERVER_API_URL}api/courses/${courseId}/exams/${examId}`;
+    }
+
+    /**
+     * Syncs the collected actions to the server.
+     * @param examActions performed actions between the last sync and now
+     * @param courseId which the exam belongs to
+     * @param examId of the current exam
+     * @param studentExamId of the student
+     * @private
+     */
+    public syncActions(examActions: ExamAction[], courseId: number, examId: number, studentExamId: number): Observable<HttpResponse<void>> {
+        const url = this.getResourceURL(courseId, examId) + `/student-exams/${studentExamId}/actions`;
+        return this.http.put<void>(url, examActions, { observe: 'response' });
+    }
```

Comment:
As there seems to be no returned data nor any header access, we don't need to observe the full response.
```suggestion
    public syncActions(examActions: ExamAction[], courseId: number, examId: number, studentExamId: number): Observable<void> {
        const url = this.getResourceURL(courseId, examId) + `/student-exams/${studentExamId}/actions`;
        return this.http.put<void>(url, examActions);
    }
```

Reply:
We need to observe the response in order to receive whether the request was successful or not :thinking: otherwise we don't know whether we can delete the actions on the client or not

Reply:
Yes i tried it (a08294dc2caca6299628181cfc059ba023673785) and reverted it again (a9d40b69ac49a0e587b79d987607c7670dca48fd) :thinking: can we check this again in an additional PR in order to not delay this :D

Reply:
Yeah it's not that important :D